File: www/src/android/AzureUpload.java
package com.azure.upload;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import android.util.Log;
import java.io.*;
import java.net.*;
import javax.net.ssl.HttpsURLConnection;

public class AzureUpload extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("uploadFiles")) {
            JSONArray fileList = args.getJSONArray(0);
            String sasToken = args.getString(1);
            String postId = args.getString(2);

            uploadFiles(fileList, sasToken, postId, callbackContext);
            return true;
        }
        return false;
    }

    private void uploadFiles(JSONArray fileList, String sasToken, String postId, CallbackContext callbackContext) {
        try {
            for (int i = 0; i < fileList.length(); i++) {
                JSONObject file = fileList.getJSONObject(i);
                String fileName = file.getString("filename");
                String fileMime = file.getString("filemime");
                byte[] fileBinary = file.getString("filebinary").getBytes();

                // Upload each file in chunks
                boolean uploadSuccess = uploadChunked(fileName, fileMime, fileBinary, sasToken);
                if (!uploadSuccess) {
                    callbackContext.error("Failed to upload file: " + fileName);
                    return;
                }
            }

            // All files uploaded successfully, call completion API
            callCompletionAPI(postId, callbackContext);

        } catch (Exception e) {
            Log.e("AzureUpload", "Upload failed: " + e.getMessage());
            callbackContext.error("Upload failed: " + e.getMessage());
        }
    }

    private boolean uploadChunked(String fileName, String fileMime, byte[] fileBinary, String sasToken) throws IOException {
        // Chunk size (e.g., 1MB)
        int chunkSize = 1024 * 1024; // 1MB
        int totalSize = fileBinary.length;
        int numChunks = (int) Math.ceil((double) totalSize / chunkSize);

        for (int i = 0; i < numChunks; i++) {
            int start = i * chunkSize;
            int end = Math.min(start + chunkSize, totalSize);
            byte[] chunk = new byte[end - start];
            System.arraycopy(fileBinary, start, chunk, 0, chunk.length);

            // Construct URL for the Azure Blob Storage SAS token
            URL url = new URL("https://yourazureblobstorageurl/" + fileName + "?sv=" + sasToken);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setRequestMethod("PUT");
            connection.setRequestProperty("x-ms-blob-type", "BlockBlob");
            connection.setDoOutput(true);

            try (OutputStream os = connection.getOutputStream()) {
                os.write(chunk);
            }

            int responseCode = connection.getResponseCode();
            if (responseCode != 201) {
                Log.e("AzureUpload", "Failed to upload chunk: " + responseCode);
                return false; // Retry logic can be implemented here
            }
        }
        return true;
    }

    private void callCompletionAPI(String postId, CallbackContext callbackContext) {
        String apiUrl = "https://personal-fjlz3d21.outsystemscloud.com/arabicschooln/rest/post/completed?id=" + postId;
        try {
            URL url = new URL(apiUrl);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            int responseCode = connection.getResponseCode();

            if (responseCode == 200) {
                callbackContext.success("All files uploaded and completion API called successfully.");
            } else {
                callbackContext.error("API call failed with response code: " + responseCode);
            }

        } catch (Exception e) {
            callbackContext.error("Error calling completion API: " + e.getMessage());
        }
    }
}

File: src/ios/AzureUpload.swift
import Foundation
import UserNotifications
import AVFoundation
import UIKit
import ImageIO
import MobileCoreServices

@objc(AzureUpload) class AzureUpload: CDVPlugin {

    @objc(uploadFiles:)
    func uploadFiles(command: CDVInvokedUrlCommand) {
        guard
            let fileList = command.arguments[0] as? [[String: Any]],
            let sasToken = command.arguments[1] as? String,
            let postId = command.arguments[2] as? String
        else {
            self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid arguments"), callbackId: command.callbackId)
            return
        }

        let totalFiles = fileList.count

        // Request notification permission
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, error in
            if granted {
                self.uploadFiles(fileList: fileList, sasToken: sasToken, postId: postId, totalFiles: totalFiles, command: command)
            } else {
                self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Notification permission not granted."), callbackId: command.callbackId)
            }
        }
    }

    func uploadFiles(fileList: [[String: Any]], sasToken: String, postId: String, totalFiles: Int, command: CDVInvokedUrlCommand) {
        DispatchQueue.global().async {
            var filesUploaded = 0

            for (index, file) in fileList.enumerated() {
                guard
                    let destination = file["destination"] as? String,
                    let fileName = file["filename"] as? String,
                    let originalName = file["originalname"] as? String,
                    let fileMime = file["filemime"] as? String,
                    let filePath = file["filepath"] as? String // Changed from fileBinary
                else {
                    self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid file data"), callbackId: command.callbackId)
                    return
                }

                let fileUri = "https://arabicschool.blob.core.windows.net/arabicschool" + destination

                self.processFile(file: file, filePath: fileUri, localFilePath: filePath, sasToken: sasToken, fileIndex: index + 1, totalFiles: totalFiles, command: command) { success in
                    if success {
                        filesUploaded += 1
                        // Check if all files have been uploaded
                        if filesUploaded == totalFiles {
                            self.callCompletionAPI(postId: postId, callbackId: command.callbackId)
                        }
                    } else {
                        self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to upload file: \(fileName)"), callbackId: command.callbackId)
                        return
                    }
                }
            }
        }
    }

    func processFile(file: [String: Any], filePath: String, localFilePath: String, sasToken: String, fileIndex: Int, totalFiles: Int, command: CDVInvokedUrlCommand, completion: @escaping (Bool) -> Void) {
        guard let mimeType = file["filemime"] as? String else {
            completion(false)
            return
        }
        
        if mimeType.starts(with: "video") {
            // Process video file
            do {
                let fileData = try Data(contentsOf: URL(fileURLWithPath: localFilePath)) // Changed from base64 decoding
                extractThumbnail(from: fileData) { thumbnailData in
                    if let thumbnailData = thumbnailData {
                        let thumbnailFilePath = filePath.replacingOccurrences(of: "video", with: "thumbnail").replacingOccurrences(of: ".mp4", with: ".jpg")
                        self.uploadChunked(filePath: thumbnailFilePath, fileBinary: thumbnailData, sasToken: sasToken, fileIndex: fileIndex, totalFiles: totalFiles, fileMime: "image/jpeg", originalName: "", postId: "", command: command) { success in
                            if success {
                                self.uploadChunked(filePath: filePath, fileBinary: fileData, sasToken: sasToken, fileIndex: fileIndex, totalFiles: totalFiles, fileMime: mimeType, originalName: "", postId: "", command: command, completion: completion)
                            } else {
                                completion(false)
                            }
                        }
                    } else {
                        self.uploadChunked(filePath: filePath, fileBinary: fileData, sasToken: sasToken, fileIndex: fileIndex, totalFiles: totalFiles, fileMime: mimeType, originalName: "", postId: "", command: command, completion: completion)
                    }
                }
            } catch {
                completion(false)
            }
        } else if mimeType.starts(with: "application/pdf") {
            // Process PDF file
            do {
                let fileData = try Data(contentsOf: URL(fileURLWithPath: localFilePath)) // Changed from base64 decoding
                self.uploadChunked(filePath: filePath, fileBinary: fileData, sasToken: sasToken, fileIndex: fileIndex, totalFiles: totalFiles, fileMime: mimeType, originalName: "", postId: "", command: command, completion: completion)
            } catch {
                completion(false)
            }
        } else {
            completion(false)
        }
    }
    
    func uploadChunked(filePath: String, fileBinary: Data, sasToken: String, fileIndex: Int, totalFiles: Int, fileMime: String, originalName: String, postId: String, command: CDVInvokedUrlCommand, completion: @escaping (Bool) -> Void) {
        let chunkSize = 1024 * 1024 // 1MB chunk size
        let totalSize = fileBinary.count
        let numChunks = Int(ceil(Double(totalSize) / Double(chunkSize)))
        var uploadedChunks = 0

        let dispatchGroup = DispatchGroup()

        for i in 0..<numChunks {
            dispatchGroup.enter()
            let start = i * chunkSize
            let end = min(start + chunkSize, totalSize)
            let chunk = fileBinary.subdata(in: start..<end)

            let uploadSuccess = self.uploadChunk(filePath: filePath, chunk: chunk, sasToken: sasToken)
            if uploadSuccess {
                uploadedChunks += 1
                let progress = Int(Double(uploadedChunks) / Double(numChunks) * 100)
                self.updateNotification(fileIndex: fileIndex, totalFiles: totalFiles, progress: progress)
            } else {
                dispatchGroup.leave()
                completion(false)
                return
            }
            dispatchGroup.leave()
        }

        dispatchGroup.notify(queue: .main) {
            self.commitUpload(postId: postId, fileMime: fileMime, originalName: originalName, fileUri: filePath, command: command, completion: completion)
        }
    }

    func uploadChunk(filePath: String, chunk: Data, sasToken: String) -> Bool {
        var request = URLRequest(url: URL(string: filePath + "?sv=" + sasToken)!)
        request.httpMethod = "PUT"
        request.setValue("BlockBlob", forHTTPHeaderField: "x-ms-blob-type")
        request.httpBody = chunk

        var success = false
        let semaphore = DispatchSemaphore(value: 0)

        let task = URLSession.shared.dataTask(with: request) { (_, response, error) in
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 {
                success = true
            } else {
                print("Failed to upload chunk: \(String(describing: error))")
            }
            semaphore.signal()
        }

        task.resume()
        semaphore.wait()

        return success
    }

    func updateNotification(fileIndex: Int, totalFiles: Int, progress: Int) {
        let content = UNMutableNotificationContent()
        content.title = "Uploading File \(fileIndex) of \(totalFiles)"
        content.body = "Upload progress: \(progress)%"
        content.sound = UNNotificationSound.default

        let request = UNNotificationRequest(identifier: "upload_progress", content: content, trigger: nil)
        UNUserNotificationCenter.current().add(request)
    }

    func commitUpload(postId: String, fileMime: String, originalName: String, fileUri: String, command: CDVInvokedUrlCommand, completion: @escaping (Bool) -> Void) {
        let url = URL(string: "https://personal-fjlz3d21.outsystemscloud.com/uploads/rest/a/commit?postid=" + postId)!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let json: [String: Any] = [
            "filemime": fileMime,
            "originalname": originalName,
            "URI": fileUri
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: json, options: [])

        let task = URLSession.shared.dataTask(with: request) { (_, response, error) in
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                completion(true)
            } else {
                print("Failed to commit file: \(String(describing: error))")
                completion(false)
            }
        }

        task.resume()
    }

    func extractThumbnail(from videoData: Data, completion: @escaping (Data?) -> Void) {
        let tempFileURL = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString + ".mp4")

        do {
            try videoData.write(to: tempFileURL)
        } catch {
            completion(nil)
            return
        }

        let asset = AVAsset(url: tempFileURL)
        let imageGenerator = AVAssetImageGenerator(asset: asset)
        imageGenerator.appliesPreferredTrackTransform = true

        let thumbnailTime = CMTime(seconds: 1, preferredTimescale: 600)
        imageGenerator.generateCGImagesAsynchronously(forTimes: [NSValue(time: thumbnailTime)]) { _, image, _, result, _ in
            if result == .succeeded, let image = image {
                let imageData = self.compressImage(UIImage(cgImage: image))
                completion(imageData)
            } else {
                completion(nil)
            }
        }
    }

    func compressImage(_ image: UIImage) -> Data? {
        return image.jpegData(compressionQuality: 0.8)
    }

    func callCompletionAPI(postId: String, callbackId: String) {
        let url = URL(string: "https://personal-fjlz3d21.outsystemscloud.com/uploads/rest/a/complete?postid=" + postId)!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let task = URLSession.shared.dataTask(with: request) { (_, response, error) in
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Upload complete"), callbackId: callbackId)
            } else {
                print("Failed to complete upload: \(String(describing: error))")
                self.commandDelegate!.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to complete upload"), callbackId: callbackId)
            }
        }

        task.resume()
    }
}
